File: ios/DeviceInfoModule.m
#import <Foundation/Foundation.h>

// RCTBridgeModule 임포트, React Native 모듈을 작성할 때 필요한 프로토콜 제공
#import "React/RCTBridgeModule.h"

// DeviceInfoModule 클래스를 NSObject 의 하위 클래스로 선언,
// RCT_EXTERN_MODULE 매크로를 사용해 React Native 모듈임을 나타냄
@interface RCT_EXTERN_MODULE(DeviceInfoModule, NSObject)

// React Native 에서 사용 가능한 메서드를 선언
// RCT_EXTERN_METHOD 매크로를 사용하며, 이 메서드는 getSystemName 이름으로 선언
// 비동기
RCT_EXTERN_METHOD(getSystemName:
                  (RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject)

// 콜백으로 처리된다
//RCT_EXTERN_METHOD(getSystemName : (RCTResponseSenderBlock)callback)

@end

File: ios/DeviceInfoModule.swift
// Foundation 프레임워크 import
import Foundation

// Objective-C에서 해당모듈을 참조할때 사용할 이름 지정
@objc(DeviceInfoModule)

// 클래스선언 및 NSObject 상속
class DeviceInfoModule : NSObject {
  
  //  Objective-C에서 해당 함수를 참조할 때 사용할 이름을 지정
  @objc
  // 문자열 "iOS"를 반환하는 함수생성
  //  콜백으로 처리
  //  func getSystemName(_ callback: RCTResponseSenderBlock) {
  //    callback(["iOS"])
  //  }

  // 비동기 처리
  func getSystemName(_ resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
        resolve("iOS")
    }
  
  @objc
  // 리액트 네이티브가 메인 스레드에서 실행되도록 설정
  static func requiresMainQueueSetup() -> Bool {
    return true;
  }
  
}
